content/shmem_alltoall_nb.tex

```diff
@@ -0,0 +1,129 @@
+\apisummary{
+  Exchanges a fixed amount of contiguous data blocks between all pairs
+  of \acp{PE} participating in the collective routine.
+}
+
+\begin{apidefinition}
+
+%% C11
+\begin{C11synopsis}
+int @\FuncDecl{shmem\_alltoall\_nb}@(shmem_team_t team, TYPE *dest, const TYPE
+*source, size_t nelems, uint32_t tag, shmem_req_h *request);
+\end{C11synopsis}
+where \TYPE{} is one of the standard \ac{RMA} types specified by Table \ref{stdrmatypes}.
+
+\begin{Csynopsis}
+\end{Csynopsis}
+\begin{CsynopsisCol}
+int @\FuncDecl{shmem\_\FuncParam{TYPENAME}\_alltoall\_nb}@(shmem_team_t team,
+TYPE *dest, const TYPE *source, size_t nelems, uint32_t tag, shmem_req_h *request);
+\end{CsynopsisCol}
+where \TYPE{} is one of the standard \ac{RMA} types and has a corresponding \TYPENAME{} specified by Table \ref{stdrmatypes}.
+
+\begin{CsynopsisCol}
+int @\FuncDecl{shmem\_alltoallmem\_nb}@(shmem_team_t team, void *dest, const
+void *source, size_t nelems, uint32_t tag, shmem_req_h *request);
+\end{CsynopsisCol}
+
+\begin{apiarguments}
+
+\apiargument{IN}{team}{A valid \openshmem team handle to a team.}%
+
+\apiargument{OUT}{dest}{Symmetric address of a data object large enough to receive
+    the combined total of \VAR{nelems} elements from each \ac{PE} in the
+    active set.
+    The type of \dest{} should match that implied in the SYNOPSIS section.}
+\apiargument{IN}{source}{Symmetric address of a data object that contains \VAR{nelems}
+    elements of data for each \ac{PE} in the active set, ordered according to
+    destination \ac{PE}.
+    The type of \source{} should match that implied in the SYNOPSIS section.}
+\apiargument{IN}{nelems}{
+  The number of elements to exchange for each \ac{PE}.
+  For \FUNC{shmem\_alltoallmem\_nb}, elements are bytes;
+  for \FUNC{shmem\_alltoall\{32,64\}\_nb}, elements are 4 or 8 bytes,
+  respectively.
+}
+\apiargument{IN}{tag}{A user defined tag to order the collective operation;
+SHMEM\_COLL\_UNORDERED can be provided if no order is required.}
+\apiargument{OUT}{request}{An opaque request handle identifying the collective
+operation.}
+
+\end{apiarguments}
+
+\apidescription{
+    The \FUNC{shmem\_alltoall\_nb} routines are collective routines. All
+    \acp{PE} in the provided team must participate in the collective. If
+    \VAR{team} compares equal to \LibConstRef{SHMEM\_TEAM\_INVALID} or is
+    otherwise invalid, the behavior is undefined.
+
+    {\bf Invocation and completion}: A call to the nonblocking alltoall routine posts the operation and returns
+    immediately without necessarily completing the operation. On the successful
+    post of the operation, an opaque request handle is created and returned. The
+    operation is completed after a call to \FUNC{shmem\_req\_test} or
+    \FUNC{shmem\_req\_wait}. When the operation is complete, the request handle
+    is deallocated and cannot be reused.
```

Comment:
To my understanding. This is currently worded in a way that both `shmem_req_test` and `shmem_req_wait` guarantee completion.

I believe the correct language exists in the intro on lines 17 and 18.

Reply:
@kwaters4 line 17 and 18 of the intro (completion semantics) states:

``
\item Completion semantics:  \openshmem programs can learn the status of the collective operations
using the \FUNC{shmem\_req\_test} routine. The operation is completed after
a call to \FUNC{shmem\_req\_test} or a call to \FUNC{shmem\_req\_wait}.
``

This is the same as here other than the ability to observe the status of the collective. I believe the text you are quoting is older.

Reply:
Maybe I misunderstand. 

I thought shmem_req_test only checks the status of the collective. but then the program continues even if the non-blocking call is incomplete. While shmem_req_wait waits and guarantees completion.

Reply:
`shmem_req_test` checks for completion and if complete, will deallocate the request handle, set the handle to SHMEM_REQ_INVALID, and returns 0.

I feel like this wording may be the confusing part:

``
The operation is completed after a call to \FUNC{shmem\_req\_test} or a call to \FUNC{shmem\_req\_wait}.
``

Maybe this wording is more clear:
``
Completion of the operation can be observed through one or more calls to \FUNC{shmem\_req\_test} or a single call to \FUNC{shmem\_req\_wait}.
``